File: Java/SelectionSort.java
// INTRODUCTION
------------------
// Merge Sort is a sorting technique which is a little more efficient than bubble sort.
// It swaps the minimum element (considering ascending order) from unsorted part and putting it at the beginning.
------------------




// TIME COMPLEXITIES
------------------------
//The time complexity of Selection sort is same for all the cases(best,worst,average)
// -- O(n^2).
------------------------






// CODE
---------------------------------------------------------------
// Java program for implementation of Selection

public class SelectionSort {
    public static void main(String[] args) {

        int [] a = { 5 ,2 , -1 , 6 , 4 , 3 , 1 , 0 };
        int n = a.length;


        for ( int i = 0 ; i < n - 1 ; i++ ){

            int minIndex = i;

            for ( int j = i + 1 ; j < n ; j++){

                if ( a[j] < a[minIndex]){
                    minIndex = j;
                }

            }
            int temp = a[i];
            a[i] = a[minIndex];
            a[minIndex] = temp;
        }





    }
}

File: Golang/SelectionSort.go
/*
Two parts are maintained - Sorted Part and Unsorted Part.
In each iteration, the minimal element is picked from the unsorted part and then is put at the end of the sorted part.
In the end, the entire array is sorted.

Time Complexity - O(n*n)

Space Complexity - O(1)
*/

package main

import "fmt"

func selectionSort(arr []int) {
	len := len(arr)
	for i := 0; i < len-1; i++ {
		minIndex := i
		for j := i + 1; j < len; j++ {
			if arr[j] < arr[minIndex] {
				arr[j], arr[minIndex] = arr[minIndex], arr[j]
			}
		}
	}
}

func main() {
	sample := []int{3, 4, 5, 2, 1, 7, 8, -1, -3}
	fmt.Println("Before sorting")
	fmt.Println(sample)
	selectionSort(sample)
	fmt.Println("After sorting")
	fmt.Println(sample)
}
